// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: shipment_status.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ShipmentStatusData with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ShipmentStatusData) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ShipmentStatusData with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ShipmentStatusDataMultiError, or nil if none found.
func (m *ShipmentStatusData) ValidateAll() error {
	return m.validate(true)
}

func (m *ShipmentStatusData) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Date

	// no validation rules for ShipmentStatusOption

	// no validation rules for Comment

	// no validation rules for CurrentLocation

	if len(errors) > 0 {
		return ShipmentStatusDataMultiError(errors)
	}

	return nil
}

// ShipmentStatusDataMultiError is an error wrapping multiple validation errors
// returned by ShipmentStatusData.ValidateAll() if the designated constraints
// aren't met.
type ShipmentStatusDataMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ShipmentStatusDataMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ShipmentStatusDataMultiError) AllErrors() []error { return m }

// ShipmentStatusDataValidationError is the validation error returned by
// ShipmentStatusData.Validate if the designated constraints aren't met.
type ShipmentStatusDataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ShipmentStatusDataValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ShipmentStatusDataValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ShipmentStatusDataValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ShipmentStatusDataValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ShipmentStatusDataValidationError) ErrorName() string {
	return "ShipmentStatusDataValidationError"
}

// Error satisfies the builtin error interface
func (e ShipmentStatusDataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sShipmentStatusData.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ShipmentStatusDataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ShipmentStatusDataValidationError{}

// Validate checks the field values on ShipmentTracking with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ShipmentTracking) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ShipmentTracking with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ShipmentTrackingMultiError, or nil if none found.
func (m *ShipmentTracking) ValidateAll() error {
	return m.validate(true)
}

func (m *ShipmentTracking) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ShipmentId

	// no validation rules for CarrierId

	// no validation rules for ShipperId

	// no validation rules for ReceiverId

	for idx, item := range m.GetShipmentStatusData() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ShipmentTrackingValidationError{
						field:  fmt.Sprintf("ShipmentStatusData[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ShipmentTrackingValidationError{
						field:  fmt.Sprintf("ShipmentStatusData[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ShipmentTrackingValidationError{
					field:  fmt.Sprintf("ShipmentStatusData[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	// no validation rules for CreatedAt

	// no validation rules for UpdatedAt

	if len(errors) > 0 {
		return ShipmentTrackingMultiError(errors)
	}

	return nil
}

// ShipmentTrackingMultiError is an error wrapping multiple validation errors
// returned by ShipmentTracking.ValidateAll() if the designated constraints
// aren't met.
type ShipmentTrackingMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ShipmentTrackingMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ShipmentTrackingMultiError) AllErrors() []error { return m }

// ShipmentTrackingValidationError is the validation error returned by
// ShipmentTracking.Validate if the designated constraints aren't met.
type ShipmentTrackingValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ShipmentTrackingValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ShipmentTrackingValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ShipmentTrackingValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ShipmentTrackingValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ShipmentTrackingValidationError) ErrorName() string { return "ShipmentTrackingValidationError" }

// Error satisfies the builtin error interface
func (e ShipmentTrackingValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sShipmentTracking.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ShipmentTrackingValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ShipmentTrackingValidationError{}
